src/query/api/v1/middleware/request_classification.go

```diff
@@ -0,0 +1,256 @@
+		return tags
+	}
+	params, err := opts.Metrics.ParseQueryParams(r, requestStart)
+	if err != nil {
+		metrics.badParams.Inc(1)
+		return tags
+	}
+
+	// NB(nate): have to check for /label/*/values this way since the URL is templated
+	labelValues := strings.Contains(path, "/label/") && strings.Contains(path, "/values")
+	if path == route.QueryRangeURL || path == route.QueryURL {
+		if params.Query == "" {
+			metrics.badParams.Inc(1)
+			return tags
+		}
+
+		if tags, err = classifyForQueryEndpoints(
+			cfg.QueryEndpointsClassification, params, w, metrics,
+		); err != nil {
+			metrics.error.Inc(1)
+			return tags
+		}
+	} else if path == route.LabelNamesURL || labelValues {
+		if tags, err = classifyForLabelEndpoints(
+			cfg.LabelEndpointsClassification, params, w, metrics,
+		); err != nil {
+			metrics.error.Inc(1)
+			return tags
+		}
+	}
+
+	return tags
+}
+
+func classifyForQueryEndpoints(
+	cfg config.QueryClassificationConfig,
+	params QueryParams,
+	w http.ResponseWriter,
+	metrics *classificationMetrics,
+) (classificationTags, error) {
+	resultsBuckets := cfg.ResultsBuckets
+	tags := newClassificationTags()
+	if len(resultsBuckets) > 0 {
+		fetchedCount := w.Header().Get(headers.FetchedSeriesCount)
+		fetched, err := strconv.Atoi(fetchedCount)
+		if err != nil {
+			return newClassificationTags(), err
+		}
+
+		tags[resultsClassification] = strconv.Itoa(resultsBuckets[0])
+		for _, bucket := range resultsBuckets {
+			if fetched >= bucket {
+				tags[resultsClassification] = strconv.Itoa(bucket)
+			}
+		}
+		metrics.classifiedResult.Inc(1)
+	}
+
+	durationBuckets := cfg.DurationBuckets
+	if len(durationBuckets) > 0 {
+		expr, err := parser.ParseExpr(params.Query)
+		if err != nil {
+			return newClassificationTags(), err
+		}
+
+		queryRange := retrieveQueryRange(expr, 0)
+		duration := params.Range()
+		totalDuration := duration + queryRange
+
+		tags[durationClassification] = durationBuckets[0].String()
+		for _, bucket := range durationBuckets {
+			if totalDuration >= bucket {
+				tags[durationClassification] = bucket.String()
+			}
+		}
+		metrics.classifiedDuration.Inc(1)
+	}
+
+	return tags, nil
+}
+
+func classifyForLabelEndpoints(
+	cfg config.QueryClassificationConfig,
+	params QueryParams,
+	w http.ResponseWriter,
+	metrics *classificationMetrics,
+) (classificationTags, error) {
+	resultsBuckets := cfg.ResultsBuckets
+	tags := newClassificationTags()
+	if len(resultsBuckets) > 0 {
+		fetchedCount := w.Header().Get(headers.FetchedMetadataCount)
+		fetched, err := strconv.Atoi(fetchedCount)
+		if err != nil {
+			return newClassificationTags(), err
+		}
+
+		tags[resultsClassification] = strconv.Itoa(resultsBuckets[0])
+		for _, bucket := range resultsBuckets {
+			if fetched >= bucket {
+				tags[resultsClassification] = strconv.Itoa(bucket)
+			}
+		}
+		metrics.classifiedResult.Inc(1)
+	}
+
+	durationBuckets := cfg.DurationBuckets
+	if len(durationBuckets) > 0 {
+		duration := params.Range()
```

Comment:
I think default (no range) should be the largest bucket fwiw, since the API treats the request as for all time